File: 202-happy-number/happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        """
        Determines if a number is a "happy number" using Floyd's Cycle-Finding Algorithm.
        All logic is inlined within this single function.

        Args:
            n: A positive integer.

        Returns:
            True if n is a happy number, and False if not.
        """
        slow = n
        fast = n

        # A do-while loop is simulated since the first iteration must always run.
        while True:
            # Advance slow pointer by one step
            s_slow = 0
            temp_slow = slow
            while temp_slow > 0:
                temp_slow, digit = divmod(temp_slow, 10)
                s_slow += digit * digit
            slow = s_slow

            # Advance fast pointer by two steps
            for _ in range(2):
                s_fast = 0
                temp_fast = fast
                while temp_fast > 0:
                    temp_fast, digit = divmod(temp_fast, 10)
                    s_fast += digit * digit
                fast = s_fast
            
            # If pointers meet, a cycle is found. Break the loop.
            if slow == fast:
                break
        
        # If the cycle is at 1, the number is happy.
        return slow == 1

File: 202-happy-number/happy-number.java
class Solution {
    /**
     * Determines if a number is a "happy number" using Floyd's Cycle-Finding Algorithm.
     * All logic is inlined within this single function.
     * @param n A positive integer.
     * @return True if n is a happy number, and false if not.
     */
    public boolean isHappy(int n) {
        int slow = n;
        int fast = n;

        do {
            // Advance slow pointer by one step
            int s_slow = 0;
            int temp_slow = slow;
            while (temp_slow > 0) {
                int digit = temp_slow % 10;
                s_slow += digit * digit;
                temp_slow /= 10;
            }
            slow = s_slow;

            // Advance fast pointer by two steps
            for (int i = 0; i < 2; i++) {
                int s_fast = 0;
                int temp_fast = fast;
                while (temp_fast > 0) {
                    int digit = temp_fast % 10;
                    s_fast += digit * digit;
                    temp_fast /= 10;
                }
                fast = s_fast;
            }
        } while (slow != fast);

        // If the loop terminated because the pointers met, check if they met at 1.
        return slow == 1;
    }
}
